scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,416 @@
+
+        # nodes have shape (input_dim, eval_points)
+        nodes = (self.nodes + 1) * (lengths / 2) + a
+        error_nodes = (self.error_nodes + 1) * (lengths / 2) + a
+
+        # Also need to multiply the weights by a scale factor equal to the determinant
+        # of the Jacobian for this coordinate change.
+        weight_scale_factor = math.prod(lengths / 2)
+
+        # weights have shape (eval_points,)
+        weights = self.weights * weight_scale_factor
+        error_weights = self.error_weights * weight_scale_factor
+
+        # f(nodes) will have shape (eval_points, output_dim)
+        # integral_estimate should have shape (output_dim)
+        integral_estimate = np.sum(weights * f(nodes, *args), axis=-1)
+
+        error_estimate = abs(np.sum(
+            error_weights * f(error_nodes, *args),
+            axis=-1
+        ))
+
+        return integral_estimate, error_estimate
+
+
+# For now, assuming that the underlying quadrature rules all have linear error
+# estimators so that we can combine the error nodes and error weights
+class ProductRule(CubatureRuleLinearError):
+    def __init__(self, base_rules):
+        self.base_rules = base_rules
+
+    @functools.cached_property
+    def nodes(self):
+        return _cartesian_product([rule.nodes for rule in self.base_rules])
+
+    @functools.cached_property
+    def error_nodes(self):
+        return _cartesian_product([rule.error_nodes for rule in self.base_rules])
+
+    @functools.cached_property
+    def weights(self):
+        return np.prod(
+            _cartesian_product([rule.weights for rule in self.base_rules]),
+            axis=0
+        )
+
+    @functools.cached_property
+    def error_weights(self):
+        return np.prod(
+            _cartesian_product([rule.error_weights for rule in self.base_rules]),
+            axis=0
+        )
+
+
+def cub(f, a, b, rule, rtol=1e-05, atol=1e-08, limit=10000, args=()):
+    if limit is None:
+        limit = np.inf
+
+    est, err = rule.estimate(f, a, b, args)
+    regions = [CubatureRegion(est, err, a, b)]
+    subdivisions = 0
+
+    while _max_norm(err) > max(atol, rtol * _max_norm(est)) and subdivisions < limit:
+        region = heapq.heappop(regions)
+
+        est_k = region.estimate
+        err_k = region.error
+
+        a_k, b_k = region.a, region.b
+        m_k = (a_k + b_k) / 2
+
+        subregion_coordinates = zip(
+            _cartesian_product(np.array([a_k, m_k]).T).T,
+            _cartesian_product(np.array([m_k, b_k]).T).T
+        )
+
+        est -= est_k
+        err -= err_k
+
+        for a_k_sub, b_k_sub in subregion_coordinates:
+            est_sub, err_sub = rule.estimate(f, a_k_sub, b_k_sub, args)
+
+            est += est_sub
+            err += err_sub
+
+            new_region = CubatureRegion(est_sub, err_sub, a_k_sub, b_k_sub)
+
+            heapq.heappush(regions, new_region)
+
+        subdivisions += 1
+
+    success = _max_norm(err) < max(atol, rtol * _max_norm(est))
```

Comment:
I think you should move the `subdivisions < limit` part of the `while` to be an `if` statement that uses `break` instead. That way, if the while loop terminates normally, you know success = True. Otherwise the `if` statement can set `success = False` before breaking and you don't need to check this again.